solarkraft/src/instrument.ts

```diff
@@ -89,30 +105,107 @@ export function tlaJsonTypeOfPrimitive(v: any) {
     }
 }
 
-// Translate a Native JS value `v` into its corresponding Apalache JSON IR representation.
-export function tlaJsonOfNative(v: any): any {
+// Return true iff `name` is a valid TLA+ Name.
+// See Lamport, Specifying Systems, 15.1. The Simple Grammar
+export function isTlaName(name: string): boolean {
+    // From Lamport, Specifying Systems, 15.1. The Simple Grammar:
+    // Letter = OneOf("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+    // Numeral = OneOf("0123456789")
+    // NameChar = Letter ∪ Numeral ∪ {"_"}
+    // Name = Tok ( (NameChar* & Letter & NameChar*) \ ({"WF_", "SF_"} & NameChar+) )
+    const re = /^[a-zA-Z0-9_]*[a-zA-Z][a-zA-Z0-9_]*$/
+    return re.test(name)
+}
+
+/**
+ * Return the Apalache JSON IR representation of a native JS value `v`.
+ *
+ * If `v` is a JS array (representing a Soroban `Vec` or `enum`), it is
+ * translated into either a TLA+ sequence, or an Apalache variant, in particular:
+ *   - a TLA+ sequence, if the array is empty or of homogenous types, or if `forceVec` is set.
+ *   - an Apalache variant, in all other cases where the array is non-empty and its first element is a string.
+ *
+ * If `v` is a JS object (representing a Soroban `Bytes`, `Map` or `struct`), it is
+ * translated into either a TLA+ string, a TLA+ record, or a TLA+ function, in particular:
+ *   - a string, if the object has a `Buffer` property, that contains a zero-padded hex encoding of the payload.
+ *   - a TLA+ record, if its keys are all strings and valid TLA+ names,
+ *   - a TLA+ function, for all other keys.
+ *
+ * For example,
+ *   [ 1, 2, 3 ]  ~~>  << 1, 2, 3 >>
+ *   [ 'A', 42 ]  ~~>  Variant("A", 42)
+ *   { type: 'Buffer', data: [190, 239] }  ~~>  "beef"
+ *   { "a": 3, "b": 5 }  ~~>  [ a |-> 3, b |-> 5 ]
+ *   { "2": 3, "4": 5 }  ~~>  SetAsFun({ <<"2", 3>>, <<"4", 5>> })
+ */
+export function tlaJsonOfNative(v: any, forceVec: boolean = false): any {
     if (typeof v === 'object') {
         if (Array.isArray(v)) {
-            // a JS array / Soroban `Vec`
-            // [ 1, 2, 3 ]  ~~>  << 1, 2, 3 >>
-            return {
-                type: 'Untyped',
-                kind: 'OperEx',
-                oper: 'TUPLE',
-                args: v.map(tlaJsonOfNative),
+            // a JS array
+            if (
+                v.length == 0 ||
+                forceVec ||
+                v.every((elem) => typeof elem === typeof v[0])
+            ) {
+                // a Soroban `Vec`
+                // [ 1, 2, 3 ]  ~~>  << 1, 2, 3 >>
+                return {
+                    type: 'Untyped',
+                    kind: 'OperEx',
+                    oper: 'TUPLE',
+                    args: v.map((arg) => tlaJsonOfNative(arg)),
```

Comment:
```suggestion
                    args: v.map(tlaJsonOfNative),
```

There's a couple of these, but this is 100% just a style preference, feel free to ignore.

Reply:
Ah, I just saw the change from
```
            // a JS array / Soroban `Vec`
            // [ 1, 2, 3 ]  ~~>  << 1, 2, 3 >>
            return {
                type: 'Untyped',
                kind: 'OperEx',
                oper: 'TUPLE',
                args: v.map(tlaJsonOfNative),
```
and I missed that you added a param to the method.